[FEATURE ember-application-engines] Initial extraction of engines from applications

dgeb opened the pull request:
This initial PR for engines follows the strategy proposed in the [Engines RFC](https://github.com/emberjs/rfcs/pull/10).

It introduces a base class, `Engine`, which is wholly extracted from (and then extended by) `Application`. Similarly, `EngineInstance` has been wholly extracted from (and then extended by) `ApplicationInstance`.

The `Engine` base class is now responsible for maintaining a registry and initializers.

The `EngineInstance` base class is now responsible for maintaining an instance-level registry and container.

`Ember.Engine` is only exposed when the new `ember-application-engines` feature flag is enabled.

The APIs and behavior of `Application` and `ApplicationInstance` remain identical, as do tests and documentation.

Some tests of extracted behavior were duplicated for engines, but the original tests for applications were kept in place. Any test duplication should be revisited before the feature flag is enabled.

The primary purpose of this PR is to allow for further experimentation with engines in an addon. It will also allow behavior to be folded into ember core gradually as experimentation proceeds.

dgeb committed:
[FEATURE ember-application-engines] Initial addition of Engine + EngineInstance

dgeb committed:
[FEATURE ember-application-engines] Move initializers from Application to Engine.

Initializers and instance initializers are a common concern for
applications and engines.

dgeb committed:
[FEATURE ember-application-engines] Refactor registry and resolver concerns for engines.

Provide a spare base implementation of `buildRegistry` for `Engine`,
which is extended by `Application`.

Also, allow for a custom `Resolver` property on `Engine` (and by
extension, `Application`).

dgeb committed:
[FEATURE ember-application-engines] Refactor container / registry initialization for engine instances.

Move registry and container creation from `ApplicationInstance` to
`EngineInstance`.

dgeb committed:
[FEATURE ember-application-engines] Feature flag for Ember.Engine.

Only expose `Ember.Engine` when feature flag is enabled.

dgeb committed:
[FEATURE ember-application-engines] Tests for engines.

Replicate appropriate tests from applications. We don’t want to yet move
tests while engines are feature flagged and in flux.

dgeb committed:
[FEATURE ember-application-engines] Tests for engine instances.

Replicate appropriate tests from application instances. We don’t want to
yet move tests while engines are feature flagged and in flux.

mixonic commented on packages/ember-application/lib/system/application-instance.js:
```diff
@@ -83,24 +79,20 @@ let ApplicationInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
 
   init() {
-    this._super(...arguments);
+    let application = get(this, 'application');
```
Interacting with `this` before `super()` is not compatible with ES6 classes, fwiw. If this can be differently factored, I think that would be ideal.

dgeb replied:
Good catch, @mixonic. I was able to refactor it such that the engine instance is aware that `application` might be set instead of `base` (i.e. if `base` is not set, then `application` is checked). if others agree that `base` is general enough to specify the base class for both engine and application instances, we can consider deprecating the `application` property at some point.

kiwiupover commented:
Thanks for your work on this @dgeb Cheers

stefanpenner commented on packages/ember-application/lib/system/engine-instance.js:
```diff
@@ -0,0 +1,79 @@
+  @uses ContainerProxyMixin
+*/
+
+let EngineInstance = EmberObject.extend(RegistryProxy, ContainerProxy, {
```
`const`

stefanpenner commented on packages/ember-application/lib/system/engine-instance.js:
```diff
@@ -0,0 +1,79 @@
+
+    if (!base) {
+      base = get(this, 'application');
+      set(this, 'base', base);
```
is `base` observable or a CP, seems like it just just be a normal get/set

stefanpenner replied:
it does not appear that `base` is guaranteed to be on the instance, we should be sure the object remains shape stable, and if a constructor can set a property it always should. Reading from the prototype for a default should be considered an anti-pattern.

dgeb replied:
I agree this is an anti-pattern that I'd like to deprecate. See this discussion with @mixonic - https://github.com/emberjs/ember.js/pull/12685#discussion-diff-46772712

I was previously setting `base` in `ApplicationInstance` before calling `_super`, but I moved away from that to be ES6 Class-compliant. Not sure which is the larger concern.

I had been setting `base` This previously

stefanpenner replied:
can we pass base in as an argument?

dgeb replied:
I can do that if I modify some `ApplicationInstance` tests, which I was trying to avoid. However, it's probably safe to do, since every real world `ApplicationInstance` will be instantiated from its associated `Application`.

dgeb replied:
Ideally we would just swap out `application` for `base` completely, which we could technically do because it's private. But I'm concerned about inadvertent breakage and think this would be better deprecated.

stefanpenner commented on packages/ember-application/lib/system/engine-instance.js:
```diff
@@ -0,0 +1,79 @@
+
+    // Create a per-instance registry that will use the application's registry
+    // as a fallback for resolving registrations.
+    let baseRegistry = get(base, '__registry__');
```
seems like it should be a normal get.

dgeb replied:
yeah, you're right - will clean up the unnecessary `get` / `set` calls

stefanpenner commented on packages/ember-application/lib/system/engine.js:
```diff
@@ -0,0 +1,422 @@
+  @uses RegistryProxy
+  @public
+*/
+let Engine = Namespace.extend(RegistryProxy, {
```
const

stefanpenner commented on packages/ember-application/lib/system/engine.js:
```diff
@@ -0,0 +1,422 @@
+      resolver: resolverFor(namespace)
+    });
+
+    registry.set = set;
```
can we avoid adding methods at runtime, instead doit at design time of the class

dgeb replied:
I just moved this code over from `Application` to here, but I agree that would be better.

stefanpenner replied:
It can happen in a later PR, I'm just noticing it because I'm reading, alternatively, `set` could be provided to the constructor, and treated as an instance value.

rwjblue replied:
Yes, I think that we should refactor to provide `set` in `options` (along with `resolver`), but that is fine for another PR...

dgeb committed:
[FEATURE ember-application-engines] Use const for module-level declarations.

dgeb committed:
[FEATURE ember-application-engines] Remove unnecessary get/set calls.

dgeb commented:
@stefanpenner thanks for the review. I've just revised as per your suggestions, with the two exceptions:

* Need to refactor `registry.set = set;`. Probably best in a separate PR since it's a change to current behavior.

* Need to consider whether to deprecate usage of the `application` property in favor of `base`. For now, the `EngineInstance` accepts either as valid, so `ApplicationInstance` does as well.

@rwjblue thoughts?

dgeb committed:
[FEATURE ember-application-engines] Expose Ember.EngineInstance and Ember.ApplicationInstance.

Expose Ember.EngineInstance and Ember.ApplicationInstance for
easy overriding. Need to reanalyze whether to continue exposing
them after feature flag is removed.

rwjblue commented on packages/ember-application/lib/index.js:
```diff
@@ -11,10 +12,19 @@ import {
+  Ember.Engine = Engine;
+
+  // Expose EngineInstance for easy overriding. Reanalyze whether to
+  // continue exposing it after feature flag is removed.
```
I agree, we should likely remove it.  `ApplicationInstance` is not exposed at all, so we should follow that (or expose it as well).

dgeb replied:
I debated exposing `ApplicationInstance` here as well - maybe we should for consistency and then remove both after the feature flag is removed?

rwjblue replied:
I don't care which way we go, but we need to be consistent.  Either both are exposed, or neither are...

rwjblue commented:
The doc strings for `Ember.Application` need to be updated to properly mark that it `@extends Ember.Engine` (other wise the docs moved simply "vanish").  Same for `Ember.EngineInstance`.

rwjblue commented on packages/ember-application/lib/system/engine.js:
```diff
@@ -0,0 +1,422 @@
+  return properties;
+}
+
+/**
```
Can you add a small description here?  This is the header page that shows up for the class in our API docs...

dgeb commented:
> The doc strings for Ember.Application need to be updated to properly mark that it @extends Ember.Engine (other wise the docs moved simply "vanish"). Same for Ember.EngineInstance.

@rwjblue Good catch. I just updated `@extends` for `Application`. I think `ApplicationInstance` and `EngineInstance` are correct though (but let me know).

rwjblue commented:
> I think ApplicationInstance and EngineInstance are correct though (but let me know).

Confirm.

dgeb commented:
@rwjblue I've added a brief description for `Engine` and exposed `Ember.ApplicationInstance` alongside `Ember.EngineInstance` (just while feature flagged).

Thanks for the review - let me know if I've overlooked any other suggestions.
